File: 54. Spiral Matrix/main.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m=len(matrix)
        n=len(matrix[0])

        GO_UP=1
        GO_DOWN=2
        GO_RIGHT=3
        GO_LEFT=4

        UP_WALL=0
        DOWN_WALL=m
        RIGHT_WALL=n
        LEFT_WALL=-1


        ans=[]

        i=0
        j=0

        DIRECTION=GO_RIGHT

        while len(ans)<m*n:
            if DIRECTION==GO_RIGHT:
                while j<RIGHT_WALL:
                    ans.append(matrix[i][j])
                    j=j+1
                DIRECTION=GO_DOWN
                RIGHT_WALL=RIGHT_WALL-1
                i=i+1
                j=j-1
            elif DIRECTION==GO_DOWN:
                while i< DOWN_WALL:
                    ans.append(matrix[i][j])
                    i=i+1
                DOWN_WALL=DOWN_WALL-1
                DIRECTION=GO_LEFT
                i=i-1
                j=j-1
            elif DIRECTION==GO_LEFT:
                while  j> LEFT_WALL:
                     ans.append(matrix[i][j])
                     j=j-1
                DIRECTION=GO_UP
                LEFT_WALL=LEFT_WALL+1
                i=i-1
                j=j+1
            else :
                while i>UP_WALL:
                    ans.append(matrix[i][j])
                    i=i-1
                UP_WALL=UP_WALL+1
                DIRECTION=GO_RIGHT
                i=i+1
                j=j+1

        return ans




        

File: 54. Spiral Matrix/main.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {

        int m = matrix.size();
        if(m==0)
            return {};
        int n = matrix[0].size();
        if(n==0)
            return {};

        

        int RIGHT=1,DOWN=2,LEFT=3,UP=4;


        std::vector<int> answers;


        int i=0;
        int j=0;


        int RIGHT_WALL=n;
        int LEFT_WALL=-1;
        int UP_WALL=0;
        int DOWN_WALL=m;


        int direction=RIGHT;
        while(answers.size() != m*n)
        {

            if(direction==RIGHT)
            {
                while(j < RIGHT_WALL)
                {
                    answers.push_back(matrix[i][j]);
                    j++;


                }

                direction=DOWN;
                RIGHT_WALL--;
                i++; j--;
            }
            else if(direction == DOWN )
            {
                while(i<DOWN_WALL)
                {
                    answers.push_back(matrix[i][j]);
                    i++;
                }

                direction=LEFT;
                DOWN_WALL--;

                i--; j--;
            }
            else if(direction==LEFT)
            {
                while(j>LEFT_WALL)
                {
                    answers.push_back(matrix[i][j]);

                    j--;
                }

                direction=UP;
                LEFT_WALL++;
                j++; i--;
            }
            else if(direction==UP)
            {
                while(i > UP_WALL)
                {
                    answers.push_back(matrix[i][j]);
                    i--;
                }
                direction=RIGHT;
                UP_WALL++;
                i++; j++;
            }

        }



    


        return answers;
    }
};